### PR TITLE
[MM-46554] Add load test functionality for global drafts

### PR DESCRIPTION
**Pull request by mylonsuren**
#### Summary

Adds load test coverage for drafts.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-46554

### CODE REVIEW

**Comment by agnivade** [comment]
@mylonsuren - just checking on the status for this. What else is pending here?

**Comment by mylonsuren** [comment]
> @mylonsuren - just checking on the status for this. What else is pending here?

@agnivade Not much, I just need to add coverage for deleting drafts and update the frequencies. Currently have some other high priority tasks in front of this, will update this PR in the coming days and get some reviews then 😄

**Comment by agarciamontoro** [comment]
Hey, @mylonsuren, anything else we can help with to get this ready for review?

File: deployment/terraform/strings.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ WorkingDirectory=/opt/mattermost
 User=ubuntu
 Group=ubuntu
 LimitNOFILE=49152
+Environment=MM_FEATUREFLAGS_GlobalDrafts
```
Comment (agnivade) [inline deployment/terraform/strings.go:20]: We need to append `=true` at the end

File: loadtest/control/simulcontroller/actions.go [hunk 2]
```diff
@@ -769,6 +769,52 @@ func (c *SimulController) createPost(u user.User) control.UserActionResponse {
...
+			fileIds <- resp.FileInfos[0].Id
+		}(filename, file.data)
+	}
+
+	wg.Wait()
+	numFiles := len(fileIds)
+	for i := 0; i < numFiles; i++ {
+		draft.FileIds = append(draft.FileIds, <-fileIds)
+	}
+
```
Comment (agnivade) [inline loadtest/control/simulcontroller/actions.go:814]: This is a complicated way to achieve this. We can just have a slice of fileIds and let each goroutine set the fileId at separate indices of that slice. That's perfectly valid and not a race condition. And then just append that slice to `draft.FileIds`

File: loadtest/control/simulcontroller/actions.go [hunk 3]
```diff
@@ -1767,6 +1813,109 @@ func (c *SimulController) updateThreadRead(u user.User) control.UserActionRespon
 	return control.UserActionResponse{Info: fmt.Sprintf("updated read state of thread %s", thread.PostId)}
 }
 
+func (c *SimulController) getDrafts(u user.User) control.UserActionResponse {
+	userId := u.Store().Id()
+
+	team, err := u.Store().CurrentTeam()
+	if errors.Is(err, memstore.ErrTeamStoreEmpty) {
```
Comment (agnivade) [inline loadtest/control/simulcontroller/actions.go:1820]: This is only returned from `Store().RandomTeam()`, so no need to check it here

File: loadtest/control/simulcontroller/actions.go [hunk 4]
```diff
@@ -1767,6 +1813,109 @@ func (c *SimulController) updateThreadRead(u user.User) control.UserActionRespon
...
+	message, err := createMessage(u, channel, false)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	draft := &model.Draft{
+		Message:   message,
+		ChannelId: channel.Id,
+		RootId:    rootId,
+		CreateAt:  time.Now().Unix() * 1000,
```
Comment (agnivade) [inline loadtest/control/simulcontroller/actions.go:1873]: Use `model.GetMillis`

File: loadtest/control/simulcontroller/actions.go [hunk 5]
```diff
@@ -1767,6 +1813,109 @@ func (c *SimulController) updateThreadRead(u user.User) control.UserActionRespon
...
+func (c *SimulController) upsertDraft(u user.User) control.UserActionResponse {
+	channel, err := u.Store().CurrentChannel()
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	var rootId = ""
+
+	// 33% of the time draft will be a thread reply
+	if rand.Float64() < 0.33 {
```
Comment (agnivade) [inline loadtest/control/simulcontroller/actions.go:1845]: How was this derived?

File: loadtest/control/simulcontroller/actions.go [hunk 6]
```diff
@@ -1767,6 +1813,109 @@ func (c *SimulController) updateThreadRead(u user.User) control.UserActionRespon
...
+}
+
+func (c *SimulController) deleteDraft(u user.User) control.UserActionResponse {
+	channel, err := u.Store().CurrentChannel()
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	team, err := u.Store().CurrentTeam()
+	if errors.Is(err, memstore.ErrTeamStoreEmpty) {
```
Comment (agnivade) [inline loadtest/control/simulcontroller/actions.go:1898]: Same as earlier

File: loadtest/control/simulcontroller/controller.go [hunk 7]
```diff
@@ -234,6 +234,18 @@ func (c *SimulController) Run() {
...
+			frequency: 0.615,
+		},
+		{
+			run:       c.getDrafts,
+			frequency: 0.555,
+		},
+		{
+			run:       c.deleteDraft,
+			frequency: 1.5,
+		},
```
Comment (agnivade) [inline loadtest/control/simulcontroller/controller.go:248]: Are all of these calculated from community metrics?

File: loadtest/control/simulcontroller/actions.go [hunk 8]
```diff
@@ -1767,6 +1813,109 @@ func (c *SimulController) updateThreadRead(u user.User) control.UserActionRespon
...
+	} else if team == nil {
+		return control.UserActionResponse{Err: control.NewUserError(errors.New("current team should be set"))}
+	}
+
+	draftId, err := u.Store().RandomDraftForTeam(team.Id)
+	if errors.Is(err, memstore.ErrDraftNotFound) {
+		return control.UserActionResponse{Info: fmt.Sprintf("no drafts found in team %v", team.Id)}
+	}
+
+	err = u.DeleteDraft(channel.Id, draftId)
```
Comment (agarciamontoro) [inline loadtest/control/simulcontroller/actions.go:1911]: Here, `channel.Id` and draftId` may be the same ID if the draft is a root post, right? I understand that is correct, but I wanted to double-check, this rootid-channelid for drafts was a bit confusing for me at first.

File: loadtest/store/memstore/store_test.go [hunk 9]
```diff
@@ -771,3 +771,85 @@ func TestThreads(t *testing.T) {
...
+		draft := &model.Draft{
+			ChannelId: channelId,
+			UserId:    user.Id,
+		}
+
+		err = s.SetDrafts(teamId, []*model.Draft{draft})
+		require.NoError(t, err)
+		draftId, err := s.RandomDraftForTeam(teamId)
+		require.NoError(t, err)
+		require.Equal(t, draft.ChannelId, draftId)
```
Comment (agarciamontoro) [inline loadtest/store/memstore/store_test.go:839]: Same comment. And here it'd actually let you to add more than one draft to check that the whole set is added.

File: loadtest/store/memstore/store_test.go [hunk 10]
```diff
@@ -771,3 +771,85 @@ func TestThreads(t *testing.T) {
...
+		draft2 := &model.Draft{
+			RootId: rootId,
+			UserId: user.Id,
+		}
+
+		err = s.SetDrafts(teamId2, []*model.Draft{draft2})
+		require.NoError(t, err)
+		draftId2, err := s.RandomDraftForTeam(teamId2)
+		require.NoError(t, err)
+		require.Equal(t, draft2.RootId, draftId2)
```
Comment (agarciamontoro) [inline loadtest/store/memstore/store_test.go:853]: And same here.

File: loadtest/control/simulcontroller/controller.go (hunk 7 above)
Comment (agarciamontoro) [reply loadtest/control/simulcontroller/controller.go:248]: Also, can we use the `minServerVersion` field in these actions? We should add that from now on to all new actions, and I want to start adding them to the current ones as well.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Thank you for this, @mylonsuren, it's great :) I left some comments to tackle, but the approach is flawless, thank you!

**Comment by agarciamontoro** [comment]
:( I will address all comments here so we can merge this PR.

**Comment by agnivade** [comment]
Sad that it's been lying for a year. Let's see if I can find some time to clean this up.

**Comment by agarciamontoro** [comment]
> Sad that it's been lying for a year. Let's see if I can find some time to clean this up.

I've been wanting to get this merged for since last year, but there was always something more urgent in the queue. Very glad if you want to tackle it!

**Comment by agnivade** [comment]
Now that I have some free time, I will take this up. It's an important part of the feature coverage. I will redo this entirely and send a fresh PR. Obviously reusing some parts of the code.

**Comment by agarciamontoro** [comment]
Thanks!

File: loadtest/control/simulcontroller/actions.go (hunk 8 above)
Comment (agnivade) [reply loadtest/control/simulcontroller/actions.go:1911]: > Here, channel.Id and draftId` may be the same ID if the draft is a root post, right?

It's the opposite. If a draft doesn't have a root post, then draftID is same as channelID.

**Pull request closed without merging**